design-proposals/klip-53-pull-queries-on-streams.md

```diff
@@ -27,17 +28,26 @@ pull-type queries.
 
 ## What is in scope
 
-* Queries should be supported on STREAM objects without “EMIT CHANGES” at the end
-* Pull queries should start at the beginning of the stream
-* Pull queries should terminate at the (current) end of the stream
+* Queries should be supported on STREAM objects without “EMIT CHANGES” at the end.
+* Pull queries should start at the beginning of the stream.
+* Pull queries should terminate. Streams do not have an "end" by definition, so we need to define
+  when the query terminates. When you issue the query, ksqlDB will scan over all the
+  data that is in the history of the stream at that moment and then terminate without
+  waiting for more data.
 * Pull queries on streams will support the same range of query operations as they
   do on tables. In other words, you will be able to filter the stream using a WHERE clause.
 
 ## What is not in scope
 
-* Pull queries on streams will not currently support grouping/aggregations.
+* Pull queries on streams will not currently support grouping/aggregations or joins.
   This would be useful but is left as future work.
-  Note that aggregations are also not currently supported on table pull queries.
+  Note that neither aggregations nor joins are currently supported on table pull queries.
+* If old records have already been dropped from the stream due to retention
+  settings, they will not be included in the results.
+* The scan over the stream may or may not pick up some new records that arrive during the query
+  execution. The only guarantee is that the query will include all records that are already in
+  the stream at the start of the query. If desired, we can make the end point stricter in the
+  future.
```

Comment:
> may or may not
Why are we fuzzy here?

> If desired, we can make the end point stricter in the future.
Why not be strict from the beginning on?

Reply:
It's because of the way the code is currently set up. All we are doing is starting up a transient query and then terminating it when it passes the endpoint, but since there are multiple partitions involved, there are actually multiple endpoints, which won't all get passed at the same time.

It would be cool to add a feature to Streams where you tell it to process up to a particular offset in each partition, so that it would be able to strictly stop processing each task at the right point, but that would turn this into a much bigger project. This way, we can release the feature sooner and then circle back later and tighten up the endpoints if we want to.

Also, it doesn't seem like strictly not including any new records that arrive after the start of the query is really needed by any known use cases. Overwhelmingly, people just want to be able to have a convenient way to inspect what is in their streams. Having strict start and end points would enable new use cases, but the use cases we have today are better served by getting the feature released asap.

Reply:
Fair enough.

I agree that we might want to extend Kafka Streams. It reminds me on the old https://cwiki.apache.org/confluence/display/KAFKA/KIP-95%3A+Incremental+Batch+Processing+for+Kafka+Streams -- the KIP tried to include the "end offset discovery" what made very complex. If we would descope the KIP and let the user provide the end-offsets, it would be much simpler.

I guess for the current use case of "scan the full topic" it might be ok to be fuzzy on the end-point... If we want to allow to do "time range scans" over a stream, if "give me all record from 10am to 1pm yesterday" we will need stricter bounds...